spec.html

```diff
@@ -28955,46 +28955,46 @@ <h1>
-            1. Else,
-              1. Let _start_ be _k_.
-              1. If _k_ + 2 ≥ _strLen_, throw a *URIError* exception.
-              1. If the code units at index (_k_ + 1) and (_k_ + 2) within _string_ do not represent hexadecimal digits, throw a *URIError* exception.
-              1. Let _B_ be the 8-bit value represented by the two hexadecimal digits at index (_k_ + 1) and (_k_ + 2).
+            1. Let _S_ be _C_.
+            1. If _C_ is the code unit 0x0025 (PERCENT SIGN), then
+              1. If _k_ + 2 ≥ _len_, throw a *URIError* exception.
+              1. Let _escape_ be the substring of _string_ from _k_ to _k_ + 3.
+              1. Let _hexDigits_ be the substring of _string_ from _k_ + 1 to _k_ + 3.
+              1. Let _parseResult_ be ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
+              1. If _parseResult_ is not a Parse Node, throw a *URIError* exception.
+              1. Let _B_ be the unsigned 8-bit value corresponding with the MV of _parseResult_.
               1. Set _k_ to _k_ + 2.
               1. Let _n_ be the number of leading 1 bits in _B_.
               1. If _n_ = 0, then
-                1. Let _C_ be the code unit whose value is _B_.
-                1. If _C_ is not in _preserveEscapeSet_, then
-                  1. Let _S_ be the String value containing only the code unit _C_.
-                1. Else,
-                  1. Let _S_ be the substring of _string_ from _start_ to _k_ + 1.
+                1. Let _c_ be the code unit whose numeric value is _B_.
+                1. If _c_ is in _preserveEscapeSet_, set _S_ to _escape_. Otherwise, set _S_ to _c_.
               1. Else,
                 1. If _n_ = 1 or _n_ > 4, throw a *URIError* exception.
-                1. If _k_ + (3 × (_n_ - 1)) ≥ _strLen_, throw a *URIError* exception.
                 1. Let _Octets_ be « _B_ ».
                 1. Let _j_ be 1.
                 1. Repeat, while _j_ &lt; _n_,
                   1. Set _k_ to _k_ + 1.
+                  1. If _k_ + 2 ≥ _len_, throw a *URIError* exception.
                   1. If the code unit at index _k_ within _string_ is not the code unit 0x0025 (PERCENT SIGN), throw a *URIError* exception.
-                  1. If the code units at index (_k_ + 1) and (_k_ + 2) within _string_ do not represent hexadecimal digits, throw a *URIError* exception.
-                  1. Let _B_ be the 8-bit value represented by the two hexadecimal digits at index (_k_ + 1) and (_k_ + 2).
-                  1. Set _k_ to _k_ + 2.
+                  1. Let _hexDigits_ be the substring of _string_ from _k_ + 1 to _k_ + 3.
+                  1. Let _parseResult_ be ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
+                  1. If _parseResult_ is not a Parse Node, throw a *URIError* exception.
+                  1. Let _B_ be the unsigned 8-bit value corresponding with the MV of _parseResult_.
```

Comment:
These re-use the aliases `_hexDigits_`, `_parseResult_`, and `_B_`, so should use `Set-to` syntax rather than `Let-be`.

Reply:
I consider these to be separate aliases in a deeper scope that shadow those in the outer scope, which has the joint benefits of a) matching likely implementation where shadowing is available, and b) allowing reuse of the pattern without gratuitous variation.

Reply:
> I consider these to be separate aliases in a deeper scope that shadow those in the outer scope,

That's reasonable from a programming language point of view, but I don't think there's anything in the spec that supports it. E.g., 5.2 "Algorithm Conventions" doesn't say anything about scopes or shadowing. It *could*, but my guess is, that would be more bother than it's worth.

Anyway, I'll leave it to the editors.

Reply:
I agree with @jmdyck; we don't currently define any scoping rules except for Abstract Closures, and it's not worth the complexity of doing so vs just picking different aliases or (my preference here) using `Set`.

Also, nested scoping would conflict with the current pattern of
```
1. If condition,
  1. Let _x_ be 1.
1. Else,
  1. Let _x_ be 2.
1. Use _x_.
```

Reply:
Sorry, I meant "deeper scope" from a _logical_ perspective rather than algorithmic. AFAICT from [Algorithm Conventions](https://tc39.es/ecma262/multipage/notational-conventions.html#sec-algorithm-conventions), `Let _x_ be …` after a preceding `Let _x_ be …` is not an error and has the effect of declaring a new alias that shadows the previous one temporally rather than lexically (i.e., "_may be referenced in any subsequent steps_"), which in the case of this algorithm are indistinguishable anyway because these declarations are in the same loop body (itself also supporting the "temporal shadowing" interpretation because each iteration encounters the same `Let` steps) and there are no references at decreasing indentation.

As noted above, I think that switching to `Set` would result in gratuitous variation for identical logic:
```diff
               1. Let _hexDigits_ be the substring of _string_ from _k_ + 1 to _k_ + 3.
               1. Let _parseResult_ be ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
               1. If _parseResult_ is not a Parse Node, throw a *URIError* exception.
               1. Let _B_ be the unsigned 8-bit value corresponding with the MV of _parseResult_.
               1. …
               1. If _n_ = 0, then
                 1. …
               1. Else,
                 1. …
                 1. Repeat, while _j_ &lt; _n_,
                   1. …
-                  1. Let _hexDigits_ be the substring of _string_ from _k_ + 1 to _k_ + 3.
-                  1. Let _parseResult_ be ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
+                  1. Set _hexDigits_ to the substring of _string_ from _k_ + 1 to _k_ + 3.
+                  1. Set _parseResult_ to ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
                   1. If _parseResult_ is not a Parse Node, throw a *URIError* exception.
-                  1. Let _B_ be the unsigned 8-bit value corresponding with the MV of _parseResult_.
+                  1. Set _B_ to the unsigned 8-bit value corresponding with the MV of _parseResult_.
```

If that is indeed the editorial preference, I think I would instead encapsulate it in a micro-operation, e.g.:
```diff
-              1. Let _hexDigits_ be the substring of _string_ from _k_ + 1 to _k_ + 3.
-              1. Let _parseResult_ be ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
-              1. If _parseResult_ is not a Parse Node, throw a *URIError* exception.
-              1. Let _B_ be the unsigned 8-bit value corresponding with the MV of _parseResult_.
+              1. Let _B_ be ReadHexOctet(_string_, _k_ + 1).
+              1. If _B_ is not an integer, throw a *URIError* exception.
               1. …
               1. If _n_ = 0, then
                 1. …
               1. Else,
                 1. …
                 1. Repeat, while _j_ &lt; _n_,
                   1. …
-                  1. Let _hexDigits_ be the substring of _string_ from _k_ + 1 to _k_ + 3.
-                  1. Let _parseResult_ be ParseText(StringToCodePoints(_hexDigits_), |HexDigits[~Sep]|).
-                  1. If _parseResult_ is not a Parse Node, throw a *URIError* exception.
-                  1. Let _B_ be the unsigned 8-bit value corresponding with the MV of _parseResult_.
+                  1. Let _continuationByte_ be ReadHexOctet(_string_, _k_ + 1).
+                  1. If _continuationByte_ is not an integer, throw a *URIError* exception.
```

Reply:
From the call today, editors are in agreement that we don't want to do this kind of shadowing. If you want to introduce a new AO here that's fine.

Reply:
Done.